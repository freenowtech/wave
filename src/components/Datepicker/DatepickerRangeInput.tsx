import { END_DATE, FirstDayOfWeek, FocusedInput, START_DATE } from '@datepicker-react/hooks';
import { compareDesc, Locale, parse, startOfDay, endOfDay } from 'date-fns';
import React, { ChangeEventHandler, MutableRefObject, useEffect, useRef, useState } from 'react';
import TetherComponent from 'react-tether';
import styled from 'styled-components';
import { compose, margin, MarginProps, width, WidthProps } from 'styled-system';
import { Colors, Elevation, MediaQueries } from '../../essentials';
import { theme } from '../../essentials/theme';
import { useGeneratedId } from '../../utils/hooks/useGeneratedId';
import { ChevronRightIcon } from '../../icons';
import { Input } from '../Input/Input';

import { Datepicker } from './Datepicker';
import { isValidDateText } from './utils/isValidDateText';
import { HelperText } from '../HelperText/HelperText';
import { dateToDisplayText } from './utils/dateToDisplayText';
import { useLocaleObject } from './utils/useLocaleObject';
import { useOnChange } from './utils/useOnChange';

type DateRangerProps = MarginProps & WidthProps;

const DateRangeWrapper = styled.div.attrs({ theme })<DateRangerProps>`
    display: inline-flex;
    align-items: center;
    position: relative;
    z-index: 0;
    width: 100%;

    input {
        &:focus,
        &:active {
            box-shadow: none;
            border-color: ${Colors.AUTHENTIC_BLUE_200};
        }
    }

    .startDate input {
        border-right: 0;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        text-align: center;
    }

    .endDate input {
        border-left: 0;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        text-align: center;
    }

    ${MediaQueries.small} {
        width: 14rem;

        .startDate input,
        .endDate input {
            text-align: left;
        }
    }

    ${compose(margin, width)}
`;

const DateArrow = styled(ChevronRightIcon)`
    position: absolute;
    left: calc(50% - 0.75rem);
    z-index: 1;
`;

const StartDateFocusedBlock = styled.div`
    background: ${Colors.ACTION_BLUE_900};
    height: 0.25rem;
    width: calc(50% - 1.5rem);

    position: absolute;
    bottom: 0;
    left: 0.75rem;
`;

const EndDateFocusedBlock = styled.div`
    background: ${Colors.ACTION_BLUE_900};
    height: 0.25rem;
    width: calc(50% - 1.5rem);

    position: absolute;
    bottom: 0;
    right: 0.75rem;
`;

export interface DateRange {
    startDate?: Date;
    endDate?: Date;
}

interface DatepickerRangeInputProps extends MarginProps, WidthProps {
    /**
     * Placeholder for start date.
     */
    startPlaceholder?: string;
    /**
     * Placeholder for end date.
     */
    endPlaceholder?: string;
    /**
     * Label for the input.
     */
    label?: string;
    /**
     * Function that used when datepicker closes without selected date.
     */
    onClose?: () => void;
    /**
     * Set the value for the date
     */
    value?: DateRange;
    /**
     * Function that used when datepicker selects new date.
     */
    onChange?: (change: DateRange) => void;
    /**
     * Minimal date to select from.
     */
    minDate?: Date;
    /**
     * Maximum date to select from.
     */
    maxDate?: Date;
    /**
     * Accepts a number for first day of the week from 0 (Sunday) to 6 (Saturday).
     * Default: 1
     */
    firstDayOfWeek?: FirstDayOfWeek;
    /**
     * Function that runs for each date and returns boolean whether date is disabled or not.
     * Default: () => false
     */
    isDateBlocked?: (date: Date) => boolean;
    /**
     * Used to align the datepicker in relation to input.
     * Default: left
     */
    placement?: 'left' | 'right' | 'center';
    /**
     * String to format dates.
     * Default: dd/MM/yyyy
     */
    displayFormat?: string;
    /**
     * String to define the locale in ISO-639-1
     * Default: en-US
     */
    locale?: string;
    /**
     * Text to be shown if error filling the input
     * or fn to be trigger as a callback when error
     */
    errorHandler?: (() => void) | string;
    /**
     * The id to be assigned to the start date input
     */
    startInputId?: string;
    /**
     * The id to be assigned to the end date input
     */
    endInputId?: string;
    /**
     * Determines the variant
     * @value `'compact'` displays only a single month
     * @default 'normal'
     */
    variant?: 'compact' | 'normal';
}

interface DateRangeInputText {
    startText: string;
    endText: string;
}

const dateRangeToDisplayText = (locale: Locale, displayFormat?: string, dateRange?: DateRange): DateRangeInputText => {
    if (!dateRange) {
        return {
            startText: '',
            endText: ''
        };
    }

    return {
        startText: dateToDisplayText(locale, displayFormat, dateRange.startDate || null),
        endText: dateToDisplayText(locale, displayFormat, dateRange.endDate || null)
    };
};

// https://date-fns.org/v2.12.0/docs/compareDesc
const isValidRange = (startDate, endDate) => {
    if (startDate && endDate) {
        return compareDesc(startDate, endDate) >= 0;
    }

    // Range remains valid until both values of the range are set
    return true;
};

const DatepickerRangeInput = ({
    minDate,
    maxDate,
    firstDayOfWeek,
    isDateBlocked,
    onClose,
    startPlaceholder,
    endPlaceholder,
    placement = 'left',
    label,
    onChange,
    displayFormat = 'dd/MM/yyyy',
    locale = 'en-US',
    value = {},
    errorHandler,
    startInputId,
    endInputId,
    variant = 'normal',
    ...rest
}: DatepickerRangeInputProps) => {
    const localeObject = useLocaleObject(locale);
    const startDateRef = useRef<HTMLInputElement>();
    const endDateRef = useRef<HTMLInputElement>();
    const [focusedInput, setFocusedInput] = useState<FocusedInput>(null);
    const [inputText, setInputText] = useState<DateRangeInputText>(
        dateRangeToDisplayText(localeObject, displayFormat, value)
    );
    const [error, setError] = useState({ startDate: false, endDate: false });
    const displayErrorMessage = typeof errorHandler === 'string';

    const startId = useGeneratedId(startInputId);
    const endId = useGeneratedId(endInputId);

    useEffect(() => {
        if (!focusedInput && (error.startDate || error.endDate) && typeof errorHandler === 'function') {
            errorHandler();
        }
    }, [error.startDate, error.endDate, focusedInput]);

    useEffect(() => {
        setInputText(dateRangeToDisplayText(localeObject, displayFormat, value));
    }, [value.startDate, value.endDate, displayFormat, localeObject]);

    useEffect(() => {
        // when the datepicker is open, bring focus to correct input

        switch (focusedInput) {
            case START_DATE:
                const startInputTarget = startDateRef.current && (startDateRef.current.children[0] as HTMLInputElement);

                if (startInputTarget) {
                    startInputTarget.focus();
                }

                break;
            case END_DATE:
                const endInputTarget = endDateRef.current && (endDateRef.current.children[0] as HTMLInputElement);

                if (endInputTarget) {
                    endInputTarget.focus();
                }

                break;
        }
    }, [focusedInput]);

    useOnChange(newFocusedInput => {
        if (!newFocusedInput && onClose) {
            onClose();
        }
    }, focusedInput);

    const handleDateChange = (startDate?: Date, endDate?: Date) => {
        if (onChange) {
            onChange({ startDate: startDate && startOfDay(startDate), endDate: endDate && endOfDay(endDate) });
        }
    };

    const handleStartDateInputChange: ChangeEventHandler<HTMLInputElement> = ({ target: { value: inputValue } }) => {
        setInputText({
            startText: inputValue,
            endText: inputText.endText
        });

        const parsedDate = inputValue === '' ? undefined : parse(inputValue, displayFormat, new Date());
        const validDate = isValidDateText(parsedDate, inputValue, displayFormat);
        const validRange = isValidRange(parsedDate, value.endDate);
        const hasError = (!!inputValue && !validDate) || !validRange;

        if (!hasError) {
            handleDateChange(parsedDate, value.endDate);
        }

        setError(state => ({
            ...state,
            startDate: hasError
        }));
    };

    const handleEndDateInputChange: ChangeEventHandler<HTMLInputElement> = ({ target: { value: inputValue } }) => {
        setInputText({
            startText: inputText.startText,
            endText: inputValue
        });

        const parsedDate = inputValue === '' ? undefined : parse(inputValue, displayFormat, new Date());
        const validDate = isValidDateText(parsedDate, inputValue, displayFormat);
        const validRange = isValidRange(value.startDate, parsedDate);
        const hasError = (!!inputValue && !validDate) || !validRange;

        if (!hasError) {
            handleDateChange(value.startDate, parsedDate);
        }

        setError(state => ({
            ...state,
            endDate: hasError
        }));
    };

    return (
        <TetherComponent
            attachment={`top ${placement}`}
            targetAttachment={`bottom ${placement}`}
            constraints={[
                {
                    to: 'window',
                    attachment: 'together'
                }
            ]}
            renderTarget={(ref: MutableRefObject<HTMLDivElement>) => (
                <>
                    <DateRangeWrapper ref={ref} {...rest}>
                        <Input
                            id={startId}
                            ref={startDateRef}
                            autoComplete="off"
                            className="startDate"
                            data-testid="start-date-input"
                            label={label}
                            placeholder={startPlaceholder}
                            onFocus={() => setFocusedInput(START_DATE)}
                            onBlur={() => setFocusedInput(null)}
                            value={inputText.startText}
                            width="100%"
                            onChange={handleStartDateInputChange}
                            data-error={error.startDate}
                        />
                        {focusedInput === START_DATE && <StartDateFocusedBlock />}
                        <DateArrow color={Colors.AUTHENTIC_BLUE_550} />
                        <Input
                            id={endId}
                            ref={endDateRef}
                            tabIndex={!inputText.startText ? -1 : 0}
                            autoComplete="off"
                            className="endDate"
                            data-testid="end-date-input"
                            placeholder={endPlaceholder}
                            onFocus={() => setFocusedInput(!value.startDate ? START_DATE : END_DATE)}
                            onBlur={() => setFocusedInput(null)}
                            value={inputText.endText}
                            onChange={handleEndDateInputChange}
                            width="100%"
                            data-error={error.endDate}
                        />
                        {focusedInput === END_DATE && <EndDateFocusedBlock />}
                    </DateRangeWrapper>
                    {displayErrorMessage && (error.startDate || error.endDate) && !focusedInput && (
                        <HelperText mt="1">{errorHandler || `error (${displayFormat})`}</HelperText>
                    )}
                </>
            )}
            renderElement={(ref: MutableRefObject<HTMLDivElement>) =>
                focusedInput && (
                    <Datepicker
                        ref={ref}
                        // TODO: refer to https://stash.intapps.it/projects/DS/repos/wave/pull-requests/104/overview?commentId=168382
                        numberOfMonths={variant === 'normal' && window.innerWidth >= 768 ? 2 : 1}
                        minBookingDays={1}
                        startDate={value.startDate || null}
                        endDate={value.endDate || null}
                        minBookingDate={minDate}
                        maxBookingDate={maxDate}
                        firstDayOfWeek={firstDayOfWeek}
                        focusedInput={focusedInput}
                        onDatesChange={({ focusedInput: focusedValue, startDate, endDate }) => {
                            setFocusedInput(focusedValue);
                            handleDateChange(startDate || undefined, endDate || undefined);
                        }}
                        isDateBlocked={isDateBlocked}
                        locale={localeObject}
                    />
                )
            }
            style={{ zIndex: Elevation.DATEPICKER }}
        />
    );
};

export { DatepickerRangeInput, DatepickerRangeInputProps };
