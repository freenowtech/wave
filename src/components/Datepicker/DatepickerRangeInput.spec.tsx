import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { format, startOfDay, endOfDay } from 'date-fns';
import userEvent from '@testing-library/user-event';
import { advanceTo, clear } from 'jest-date-mock';
import * as React from 'react';
import { DatepickerRangeInput, DatepickerRangeInputProps, DateRange } from './DatepickerRangeInput';

const ControlledRangeInput: React.FC<DatepickerRangeInputProps> = ({
    onChange,
    ...rest
}: DatepickerRangeInputProps) => {
    const [range, setRange] = React.useState<DateRange>();

    return (
        <DatepickerRangeInput
            value={range}
            onChange={(...dates) => {
                onChange?.(...dates);
                setRange(dates[0]);
            }}
            {...rest}
        />
    );
};

describe('DatepickerRangeInput', () => {
    const today = new Date(1999, 1, 2);

    beforeAll(() => {
        advanceTo(today);
    });

    afterAll(() => {
        clear();
    });

    it('renders the default props', () => {
        expect(render(<DatepickerRangeInput />).container).toMatchSnapshot();
    });

    describe('should call onClose function', () => {
        it('when clicking outside', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<DatepickerRangeInput onClose={mockCloseHandler} />);

            // Open datepicker
            userEvent.click(getByTestId('start-date-input'));

            // Click outside
            fireEvent.blur(getByTestId('start-date-input'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
        it('exactly once after non initial date selection', () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId, getAllByText } = render(<ControlledRangeInput onClose={mockCloseHandler} />);
            const selectDate = day => {
                // Open datepicker
                userEvent.click(getByTestId('start-date-input'));

                const buttons = getAllByText(day);
                // Select the first date
                act(() => userEvent.click(buttons[0]));
                // Select the second date
                act(() => userEvent.click(buttons[1]));
            };

            // initial date selection
            selectDate('01');

            mockCloseHandler.mockReset();

            selectDate('02');

            expect(mockCloseHandler).toBeCalledTimes(1);
        });
    });

    it('should call onChange function after the date selection', async () => {
        const mockChangeHandler = jest.fn();

        const { getByTestId, getAllByText } = render(<ControlledRangeInput onChange={mockChangeHandler} />);

        // Open datepicker
        userEvent.click(getByTestId('start-date-input'));

        const buttons = getAllByText('01');

        // Select the first date
        userEvent.click(buttons[0]);

        // Select the second date
        userEvent.click(buttons[1]);
        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();

        await waitFor(() =>
            expect(mockChangeHandler).toHaveBeenCalledWith({
                startDate: startOfDay(new Date(currentYear, currentMonth, 1)),
                endDate: endOfDay(new Date(currentYear, currentMonth + 1, 1))
            })
        );
    });

    it('calls onChange function after update input value', () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId } = render(<ControlledRangeInput onChange={mockChangeHandler} />);
        const startInput = getByTestId('start-date-input');
        const endInput = getByTestId('end-date-input');

        // Focus on datepicker
        userEvent.click(startInput);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedRange = {
            start: new Date(currentYear, currentMonth, 1),
            end: new Date(currentYear, currentMonth, 2)
        };
        const formattedDateStart = format(intendedRange.start, 'dd/MM/yyyy');
        const formattedDateEnd = format(intendedRange.end, 'dd/MM/yyyy');

        // Simulate an input change
        userEvent.type(startInput, formattedDateStart);
        userEvent.type(endInput, formattedDateEnd);

        expect(mockChangeHandler).toHaveBeenCalledTimes(2);
        expect(mockChangeHandler).toHaveBeenNthCalledWith(2, {
            startDate: startOfDay(intendedRange.start),
            endDate: endOfDay(intendedRange.end)
        });
    });

    it('allows to set a range with a single day', () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId } = render(<ControlledRangeInput onChange={mockChangeHandler} />);
        const startInput = getByTestId('start-date-input');
        const endInput = getByTestId('end-date-input');

        // Focus on datepicker
        userEvent.click(startInput);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedRange = {
            start: new Date(currentYear, currentMonth, 1),
            end: new Date(currentYear, currentMonth, 1)
        };
        const formattedDateStart = format(intendedRange.start, 'dd/MM/yyyy');
        const formattedDateEnd = format(intendedRange.end, 'dd/MM/yyyy');

        // Simulate an input change
        userEvent.type(startInput, formattedDateStart);
        userEvent.type(endInput, formattedDateEnd);

        expect(mockChangeHandler).toHaveBeenCalledTimes(2);
        expect(mockChangeHandler).toHaveBeenNthCalledWith(2, {
            startDate: startOfDay(intendedRange.start),
            endDate: endOfDay(intendedRange.end)
        });
    });

    it('renders an error feedback if gets filled with invalid value', () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId } = render(<ControlledRangeInput onChange={mockChangeHandler} />);
        const startInput = getByTestId('start-date-input');
        const endInput = getByTestId('end-date-input');

        // Focus on datepicker
        userEvent.click(startInput);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedRange = {
            start: new Date(currentYear, currentMonth, 2),
            end: new Date(currentYear, currentMonth, 1)
        };
        const formattedDateStart = format(intendedRange.start, 'dd/MM/yyyy');
        const formattedDateEnd = format(intendedRange.end, 'dd/MM/yyyy');

        // Simulate an input change
        userEvent.type(startInput, formattedDateStart);
        userEvent.type(endInput, formattedDateEnd);

        expect(startInput).toHaveAttribute('data-error', 'false');
        expect(endInput).toHaveAttribute('data-error', 'true');
        // Cause when error is thrown there is no actual parsing of date
        expect(mockChangeHandler).toHaveBeenCalledTimes(1);
    });

    it('on error calls errorHandler as a fn if a callback is provided', () => {
        const mockChangeHandler = jest.fn();
        const mockErrorTextHandler = jest.fn();
        const { getByTestId } = render(
            <ControlledRangeInput onChange={mockChangeHandler} errorHandler={mockErrorTextHandler} />
        );
        const startInput = getByTestId('start-date-input');
        const endInput = getByTestId('end-date-input');

        // Focus on datepicker
        userEvent.click(startInput);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedRange = {
            start: new Date(currentYear, currentMonth, 2),
            end: new Date(currentYear, currentMonth, 1)
        };
        const formattedDateStart = format(intendedRange.start, 'dd/MM/yyyy');
        const formattedDateEnd = format(intendedRange.end, 'dd/MM/yyyy');

        // Simulate an input change
        userEvent.type(startInput, formattedDateStart);
        userEvent.type(endInput, formattedDateEnd);

        // Error is explicitly called after blur the input so we avoid calls while user type
        fireEvent.blur(startInput);

        expect(mockErrorTextHandler).toHaveBeenCalledTimes(1);
    });

    it('closes the calendar of the first datepicker and opens the calendar of the second', () => {
        const { getAllByTestId, getAllByText } = render(
            <>
                <DatepickerRangeInput />
                <DatepickerRangeInput />
            </>
        );

        const inputs = getAllByTestId('start-date-input');
        expect(inputs).toHaveLength(2);

        userEvent.click(inputs[0]);
        expect(getAllByText('01')).toHaveLength(2);

        userEvent.click(inputs[1]);
        expect(getAllByText('01')).toHaveLength(2);
    });
});
