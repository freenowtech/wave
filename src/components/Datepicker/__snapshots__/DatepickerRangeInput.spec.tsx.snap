// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatepickerRangeInput renders the default props 1`] = `
.c2 {
  margin: 0;
  box-sizing: border-box;
  background: #FFFFFF;
  border-radius: 0;
  color: #001E3E;
  font-size: 1rem;
  font-family: "Open Sans",sans-serif;
  -webkit-transition: box-shadow 100ms,border 100ms;
  transition: box-shadow 100ms,border 100ms;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  border-radius: 0.25rem;
  border: 0.0625rem solid #C6CDD4;
  font-size: 1rem;
  height: 3rem;
  padding: 0 0.75rem;
}

.c2::-webkit-input-placeholder {
  color: #9CA7B4;
}

.c2::-moz-placeholder {
  color: #9CA7B4;
}

.c2:-ms-input-placeholder {
  color: #9CA7B4;
}

.c2::placeholder {
  color: #9CA7B4;
}

.c2:active,
.c2:focus {
  border-color: #096BDB;
  box-shadow: inset 0 0 0 0.0625rem #096BDB;
}

.c2:disabled {
  color: #C6CDD4;
  border-color: #C6CDD4;
  box-shadow: none;
  cursor: not-allowed;
}

.c2:disabled::-webkit-input-placeholder {
  color: #C6CDD4;
}

.c2:disabled::-moz-placeholder {
  color: #C6CDD4;
}

.c2:disabled:-ms-input-placeholder {
  color: #C6CDD4;
}

.c2:disabled::placeholder {
  color: #C6CDD4;
}

.c2:-webkit-autofill,
.c2:-webkit-autofill:hover,
.c2:-webkit-autofill:focus,
.c2:-webkit-autofill:active {
  -webkit-text-fill-color: #001E3E;
  -webkit-transition: background-color 99999999ms ease 99999999ms;
  transition: background-color 99999999ms ease 99999999ms;
}

.c2 + .sc-AxgMl {
  color: #9CA7B4;
  background: #FFFFFF;
  background: linear-gradient(0deg,#FFFFFF calc(50% + 0.0625rem),transparent 50%);
}

.c2:disabled + .sc-AxgMl {
  color: #C6CDD4;
}

.c2:-webkit-autofill + .sc-AxgMl,
.c2:-webkit-autofill:hover + .sc-AxgMl,
.c2:-webkit-autofill:focus + .sc-AxgMl,
.c2:-webkit-autofill:active + .sc-AxgMl {
  font-weight: 600;
  top: -0.5rem;
  font-size: 0.75rem;
}

.c2:focus:not(:disabled) + .sc-AxgMl {
  font-weight: 600;
  top: -0.5rem;
  font-size: 0.75rem;
  color: #096BDB;
  background: #FFFFFF;
  background: linear-gradient(0deg,#FFFFFF calc(50% + 0.0625rem),transparent 50%);
}

.c1 {
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  width: 100%;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: 0;
  width: 100%;
}

.c0 input:focus,
.c0 input:active {
  box-shadow: none;
  border-color: #C6CDD4;
}

.c0 .startDate input {
  border-right: 0;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  text-align: center;
}

.c0 .endDate input {
  border-left: 0;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  text-align: center;
}

.c3 {
  position: absolute;
  left: calc(50% - 0.75rem);
  z-index: 1;
}

@media screen and (min-width:30rem) {
  .c0 {
    width: 14rem;
  }

  .c0 .startDate input,
  .c0 .endDate input {
    text-align: left;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 startDate"
      width="100%"
    >
      <input
        autocomplete="off"
        class="sc-AxirZ c2"
        data-error="false"
        data-testid="start-date-input"
        type="text"
        value=""
      />
    </div>
    <svg
      class="c3"
      color="#637689"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M13.675 12L7.5 5h2.676l6.222 7-6.222 7H7.5z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    <div
      class="c1 endDate"
      width="100%"
    >
      <input
        autocomplete="off"
        class="sc-AxirZ c2"
        data-error="false"
        data-testid="end-date-input"
        tabindex="-1"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
